src/setup_runs/config_read_functions.py

```diff
@@ -0,0 +1,330 @@
+    sst_monthly_dir: str
+    """directory containing monthly SST files"""
+    sst_daily_dir: str
+    """directory containing daily SST files"""
+    sst_monthly_pattern: str
+    """pattern for matching monthly SST files (date-time substitutions recognised)"""
+    sst_daily_pattern: str
+    """pattern for matching daily SST files (date-time substitutions recognised)"""
+    sst_vtable: str
+    """VTable file for the SST GRIB files"""
+    analysis_pattern_upper: str
+    """pattern for matching the full-atmosphere analysis files (date-time substitutions recognised as well as shell wildcards)"""
+    analysis_pattern_surface: str
+    """pattern for matching the surface-level analysis files (date-time substitutions recognised as well as shell wildcards)"""
+    analysis_vtable: str
+    """VTable file for the SST analysis files"""
+    wrf_run_dir: str
+    """directory containing WRF input tables and data-files"""
+    wrf_run_tables_pattern: str
+    """pattern to match to get the WRF input tables and data-files (within the folder ${wrf_run_dir}"""
+
+
+def read_config_file(configFile: str) -> str :
+    """
+    Read and return the content of a configuration file.
+
+    Parameters
+    ----------
+    configFile
+        The path to the configuration file.
+
+    Returns
+    -------
+        The content of the configuration file.
+
+    """
+
+    assert os.path.exists(
+        configFile
+    ), f"No configuration file was found at {configFile}"
+
+    try :
+        with open(configFile, 'rt') as f :
+            input_str = f.read()
+        return input_str
+    except Exception as e :
+        print("Problem reading in configuration file")
+        print(e)
+        sys.exit()
+
+
+def parse_config(input_str: str) -> dict[str, str | bool | int] :
+    """
+    Parse the input string  and remove comments.
+
+    Parameters
+    ----------
+    input_str
+        The input string containing configuration data.
+
+    Returns
+    -------
+        The parsed configuration data.
+    """
+
+    try :
+        # strip out the comments
+        input_str = re.sub(r'#.*\n', '\n', input_str)
+        return json.loads(input_str)
+    except Exception as e :
+        print("Problem parsing in configuration file")
+        print(e)
+        sys.exit()
+
+
+def add_environment_variables(config: dict[str, str | bool | int], environmental_variables: dict[str, str]) -> dict[
+    str, str | bool | int] :
+    """
+    Add environment variables to the configuration that may be needed for substitutions.
+
+    Parameters
+    ----------
+    config
+        The configuration dictionary.
+
+    Returns
+    -------
+        The updated configuration dictionary with added environment variables.
+
+    """
+    envVarsToInclude = config["environment_variables_for_substitutions"].split(',')
+
+    for envVarToInclude in envVarsToInclude :
+        config[envVarToInclude] = environmental_variables[envVarToInclude]
```

Comment:
Instead of only adding variables that can be found in the environment variables (see old code below), all variables specified under `config[“environment_variables_for_substitutions”]` must now be found in the environment variables. If a variable cannot be added, an error will occur (instead of later during substitution). 

```python
## add some environment variables to the config that may be needed for substitutions
envVarsToInclude = config["environment_variables_for_substitutions"].split(',')
for envVarToInclude in envVarsToInclude:
    if envVarToInclude in list(os.environ.keys()):
        config[envVarToInclude] = os.environ[envVarToInclude]

```